jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs

```diff
@@ -9,32 +9,118 @@ mod error;
 use std::sync::Arc;
+use std::thread::sleep;
+use std::time::Duration;
 
+/// A wrapper around `reqwest::blocking::Client` providing HTTP request functionality with retry logic.
+///
+/// The `HttpClient` struct allows for sending GET requests with a retry mechanism that attempts to
+/// fetch the requested resource up to a maximum number of times if an error occurs.
+pub struct HttpClient {
+    client: reqwest::blocking::Client,
+    max_retries: u32,
+    retry_delay: Duration,
+}
+
+impl HttpClient {
+    pub fn default() -> Result<Self, KeyServiceError> {
+        let client = Client::builder()
+            .build()
+            .map_err(KeyServiceError::HttpClientInitialization)?;
+        Ok(Self {
+            client,
+            max_retries: 3,
+            retry_delay: Duration::from_secs(1),
+        })
+    }
+
+    #[cfg(test)]
+    pub fn fast_retry_delay() -> Result<Self, KeyServiceError> {
+        let client = Client::builder()
+            .build()
+            .map_err(KeyServiceError::HttpClientInitialization)?;
+        Ok(Self {
+            client,
+            max_retries: 3,
+            retry_delay: Duration::from_millis(10),
+        })
+    }
+
+    /// Sends a GET request to the specified URI with retry logic.
+    ///
+    /// This method will attempt to fetch the resource up to 3 times, with an increasing delay
+    /// between each attempt.
+    fn get(&self, uri: &str) -> Result<reqwest::blocking::Response, KeyServiceError> {
+        // Fetch the JWKS from the jwks_uri
+        let mut attempts = 0;
+        let response = loop {
+            match self.client.get(uri).send() {
+                Ok(response) => break response, // Exit loop on success
+                Err(e) if attempts < self.max_retries => {
+                    attempts += 1;
+                    // TODO: pass this message to the logger
+                    eprintln!(
+                        "Request failed (attempt {} of {}): {}. Retrying...",
+                        attempts, self.max_retries, e
+                    );
+                    sleep(self.retry_delay * attempts);
+                },
+                Err(e) => return Err(KeyServiceError::MaxHttpRetriesReached(e)), // Exit if max retries exceeded
+            }
+        };
+        response
+            .error_for_status()
+            .map_err(KeyServiceError::HttpStatus)
+    }
+}
+
+/// A service that manages key retrieval and caching for OpenID Connect configurations.
 /// Retrieves a [`DecodingKey`]-s based on the provided jwks_uri.
 fn fetch_decoding_keys(
     jwks_uri: &str,
-    http_client: &Client,
+    http_client: &HttpClient,
 ) -> Result<HashMap<Box<str>, DecodingKey>, KeyServiceError> {
-    let jwks: JwkSet = http_client
-        .get(jwks_uri)
-        .send()
-        .map_err(KeyServiceError::Http)?
-        .error_for_status()
-        .map_err(KeyServiceError::Http)?
+    let result = http_client.get(jwks_uri)?;
+    println!("fetching jwks");
```

Comment:
throw the logger through the config in `ServiceFactory` and use it to log...

Reply:
i removed the print statements here: [dde45a6](https://github.com/JanssenProject/jans/pull/9982/commits/dde45a6e5a07038da1d8028d281558cbb7cc1199).
I will just implement the logger in a separate issue since it requires a huge change due to how the logger is initialized and this feels out of scope.